Add lidarr as downloader

avedor opened the pull request:
Implements an MVP version of a Lidarr downloader using the [lidarr-go sdk](https://github.com/devopsarr/lidarr-go). Currently, can't seem to automatically download using slskd, but the general workflow is functional. Would probably close #16

avedor committed:
Implement initial support for Lidarr downloader

Add helper funcs; implement hour cooldown for search

Check for rejected releases

Add cleanup func and worker

Add check to artist adding

Mark track as present

avedor committed:
Implement initial support for Lidarr downloader

Add helper funcs; implement hour cooldown for search

Check for rejected releases

Add cleanup func and worker

Add check to artist adding

Mark track as present

avedor commented:
sounds good! i've got the actual logic written up. how would you envision something like that running? i've added it as a goroutine for now, but if there's a different architecture you'd prefer, just name it!

LumePart commented:
> i've added it as a goroutine for now

Yeah, goroutine is what I was thinking too

I looked through and ran the code and I'm a bit confused about findArtist() and addArtistIfNeeded(), previously when I investigated Lidarr, I did a manual flow of API requests to download the album and I only used ``/api/v1/album``[ endpoint](https://lidarr.audio/docs/api/#/Album/post_api_v1_album).

This is the request I used for adding and sending the search to indexers:

POST http://URL:PORT/api/v1/album
**Header**: X-Api-Key: API_KEY
**Body**: 

```json
{
 "foreignAlbumId": "70bdc19f-66fb-4adf-9124-925ace531101",
 "monitored": true,
 "qualityProfileId": 1,
 "rootFolderPath": "/music",
 "monitored": true,
 "artist": {
     "foreignArtistId": "89ea85c2-c90a-46ea-93db-9bed95d5c8f8",
     "qualityProfileId": 1,         // Artist's Quality Profile ID
     "metadataProfileId": 2,        // Artist's Metadata Profile ID
     "monitored": true,
     "rootFolderPath": "/music"
     },
 "addOptions": {
     "searchForNewAlbum": true
     }
}
```
AlbumID and artistID came from [album lookup](https://lidarr.audio/docs/api/#/AlbumLookup/get_api_v1_album_lookup) and the ProfileID's could probably be config parameters if the SDK supports it
I had trouble implementing the same flow with the SDK, but I can try again tomorrow.

And i fixed a small issue with track.MainArtist in the dev branch, when searching for albums it's best to use that field, track.Artist could contain all the artists on the track (i.e Justice feat. Tame Impala)

avedor commented:
Yeah! so the whole flow was a little confusing for me as well and the sdk doesn't seem to map 1:1 onto the api endpoints which is... frustrating. 

my understanding is we've got the foreign (read musicbrainz) ids for everything and the internal (lidarr) ids. Lidarr ids for artist/album don't get set if the artist isn't in lidarr already, so `findArtist` does the lookup using a string query--though could probably be switched to use the artist mbid if we add that as a field in the track struct--then the `addArtistIfNeeded` is creating the artist in lidarr, if it isn't already there, using defaults pulled from the root settings.

i'll dig into the sdk a little further to see if i can bypass a lot of the lookups and just use the mbids to create the album in lidarr

avedor committed:
Use main artist in search

avedor committed:
Use main artist in search

avedor commented:
just for complete context, between these changes, and those in #22, i've now got a new playlist in jellyfin with 33/50 tracks!

as far as the flow using sdk is concerned, i think we're roughly stuck at:

1. look up artist using string term search and return `ArtistResource`
2. pass that into `addArtistIfNeeded` which checks to see if a. the artist has a path set and b. has a non-zero timestamp for when the artist was added. if not, it creates a new `ArtistResource` with some defaults from the root settings and calls `CreateArtist` to add the artist to lidarr
3. repeat the same basic flow for album and release

LumePart commented:
Wouldn't it be more straightforward to do it without the SDK? It would reduce the amount of flows needed and would give us greater control of the request payloads. Especially when we don't need most of the API functionality (most likely 4-5 endpoints for a full workflow).

The flow would look pretty much like this:
1. Query album to get artistID and AlbumID
2. Request album to be downloaded/sent to indexer
3. Periodically check if it's downloading
4. If downloaded, mark track as Present otherwise clean Lidarr of artist/track

Feel free to debate me on this, maybe i just don't get the point of the SDK when Lidarr's API documentation is pretty good.

avedor committed:
Use raw api calls

Add download monitor

Add album release to request

More updates to lidarr

Don't monitor the whole artist

Fix rebase

More rebase fixes

Remove timeout from main.go

Fix rebase more

Fix rebase more

avedor committed:
Use raw api calls

Add download monitor

avedor commented:
This is the first time I've used an SDK, so I'm happy to do it either way! I just pushed up a couple squashed commits that should do all of this a little more cleanly (storing the MBIDs) in the track struct for ease. I also updated the goroutine to both clear stale downloads and mark present newly available tracks

Looking forward to your thoughts!

avedor committed:
Add album release to request

More updates to lidarr

avedor committed:
Don't monitor the whole artist

Fix rebase

More rebase fixes

LumePart commented:
Made a few changes in the dev branch, main changes include:
- MonitorDownloads([]*models.Track) function to the Downloader interface, Lidarr most likely can make use of this.
- queryTrack() and getTrack() are asynchronous now
- A few fields to Track struct, including a field to store the track size (if needed) and an field from ListenBrainz (length (in ms))

I'm also pretty much finished with implementing slskd, you can check how I [implemented download monitoring there](https://github.com/LumePart/Explo/blob/feat/slskd/src/downloader/slskd.go)

avedor committed:
Remove timeout from main.go

Fix rebase more

Fix rebase more

avedor committed:
DRY downloader checker

Fix post rebase

Remove redeclarations

Update sample env; reorganize

Add PUID/GID support to Docker container

Fixup rebase

Update lidarr monitoring

Fix config

Fix ReadEnv rebase

Fixup rebase

Restore a couple files

Unformat config

Ignore vscode

Restore downloader

Add fields to type

Make lidarr timeout an int

Fix env reading

Remove scheme

Fix cleanup tasks

Use track title

Structure logging

Fixup rebase

Cut down on changes

Fix linting issues

avedor committed:
DRY downloader checker

Fix post rebase

Remove redeclarations

tessamerrill commented:
>  Is Lidarr fixed now? I heard there were some issues with their metadata servers, but some people say it's working for them 🤷

I'll jump in and say that I've been using a [fork](https://github.com/blampe/hearring-aid) that uses a different metadata server (that actually works).

tessamerrill commented:
> I think it would be better to move the monitor refactoring into a separate pull request, after Lidarr is fully implemented. This would simplify testing and allow for a cleaner refactor, especially since monitor.go will likely require a shared struct.
> 
> 
> 
> Right now, I can't test Lidarr properly because my test environment doesn't like hearring-aid (the MusicBrainz server requirements are a bit too much for it). Since I'm not sure what fields Lidarr needs yet, I can't provide meaningful feedback at the moment.
> 
> 
> 
> But according to Lidarr's Issues page, they are pretty close in making it work again.

I would be more than happy to test using my environment, are there any docker images or documentation i can use?

avedor commented:
So i've moved the refactor onto its own branch (#40). It may make more sense to address that first so that this PR can leverage that function instead of implementing it again and removing it in the follow up. Everything else, I think depends on that?

avedor committed:
Update sample env; reorganize

Add PUID/GID support to Docker container

Fixup rebase

avedor committed:
Update lidarr monitoring

Fix config

Fix ReadEnv rebase

Fixup rebase

Restore a couple files

Unformat config

Ignore vscode

Restore downloader

Add fields to type

Make lidarr timeout an int

Fix env reading

Remove scheme

Fix cleanup tasks

avedor commented:
I appreciate the update! I think I've gotten everything hooked up correctly, but there's been an unfortunate wrinkle in that the Lidarr metadata server is having issues. As it stands, the only response I can seemingly get back from the Lidarr API is a 503. 

It looks like they're nearly done with their upgrade/transition, but until it's working even remotely reliably, there's not a whole lot more testing I'll be able to do. :slightly_frowning_face:

LumePart commented:
Thanks!
Seems like fixing the metadata server is taking a lot longer than anticipated.

I have a question about the `LIDARR_SCHEME` variable, it's still used in some of the functions, but not in the Monitor functions. Can it be taken out completely? Seems redundant

avedor committed:
Use track title

Structure logging

Permanently commented:
> Currently, can't seem to automatically download using slskd

If it's added to Lidarr then [Soularr](https://github.com/mrusse/soularr) can do the rest, no?

jasontucker commented:
I'd love to see this happen. I've been using this for a week now and im really enjoying it but not a fan of it downloading from YouTube when I have lidar right there that could be used for the downloading.

avedor committed:
Update gui to include lidarr

Build test branch

Update condition

Fix request calls

Fix the linter

Simplify api calls

Debug

Try to search on add

Remove old structs; update querytrack

Debug and comment out helper func

More debugging

Add new helper func

Fix call; add debug

Use release group as id

Debug

More debugging

Add release group is helper

Get release group

avedor committed:
Don't add an album that exists

Remove debugs

Set trackfile after download starts

Fix rebase

Remove extraneous code

Restore untouched file

LumePart committed:
check trackID when monitoring; check Lidarr history for completed downloads

LumePart committed:
use queueID when deleting lidarr downloads

LumePart committed:
implement checking search status

LumePart committed:
change lidarr config values

LumePart committed:
use helper; handle edge-cases

LumePart committed:
add pagination and some fixes

LumePart committed:
replace print with slog

LumePart committed:
add rootfolder search by name

dammitjeff committed:
wire LidarrURL + LidarrAPIKey to wizard handler, persist MBID's through playlist cache

dammitjeff committed:
add ForeignArtistID to Album struct and update getReleaseGroupId logic

LumePart committed:
save root dir when initializing lidarr

LumePart committed:
Track ID based lidarr monitoring

LumePart committed:
create album cache

LumePart committed:
cache tracks from lidarr response

LumePart committed:
split GetTrack to subfunctions

LumePart committed:
check cache before searching track

LumePart committed:
add pointe

LumePart committed:
move MoveDownload under monitor struct

LumePart committed:
remove MoveDownload from downloader

LumePart committed:
remove comparison

LumePart committed:
create albumcache on init

LumePart committed:
add post search album filtering

LumePart committed:
fall back to first available rootfolder, when not found by name

LumePart commented:
I'm ready to merge this to dev soon (possibly tomorrow or Sunday).
* MoveDownloads has been implemented
* Added a max duration variable to monitoring, so all tracks/albums not downloaded within X minutes (default 120) will get discarded and other downloaders can try getting the track.
* Added a few fields to the Wizard
